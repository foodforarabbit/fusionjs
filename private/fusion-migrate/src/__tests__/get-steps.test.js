const fs = require('fs');
const path = require('path');
const babel = require('@babel/core');
const format = require('../utils/codemod-test.js').format;
const getSteps = require('../get-steps.js');
const parserOpts = require('../parser-opts.js');
const modDataDependency = require('../codemods/data-dependency/plugin.js');
const modRedirect = require('../codemods/mod-redirect/plugin.js');
const modReactRouter = require('../codemods/mod-react-router-4/plugin.js');
const modIndexRedirect = require('../codemods/mod-index-redirect/plugin.js');
const modIndexRoute = require('../codemods/mod-index-route/plugin.js');
const modHoistRoutes = require('../codemods/hoist-routes/plugin.js');
const modReplaceRouterImports = require('../codemods/replace-react-router-imports/plugin.js');

test('getSteps 14', async () => {
  const steps = getSteps({version: 14});
  expect(Array.isArray(steps)).toEqual(true);
  steps.forEach(s => {
    if (s.id) {
      expect(typeof s.id).toEqual('string');
    } else {
      expect(typeof s.name).toEqual('string');
    }
    expect(typeof s.step).toEqual('function');
  });
  expect(steps.length).toBeGreaterThan(0);
});

test('getSteps 13', async () => {
  const steps = getSteps({version: 13});
  expect(Array.isArray(steps)).toEqual(true);
  steps.forEach(s => {
    if (s.id) {
      expect(typeof s.id).toEqual('string');
    } else {
      expect(typeof s.name).toEqual('string');
    }
    expect(typeof s.step).toEqual('function');
  });
  expect(steps.length).toBeGreaterThan(0);
});

const expectedRoutingSteps = {
  'mod-data-dependency': modDataDependency,
  'mod-redirect': modRedirect,
  'mod-index-redirect': modIndexRedirect,
  'mod-index-route': modIndexRoute,
  'mod-react-router': modReactRouter,
  'mod-hoist-routes': modHoistRoutes(),
  'mod-replace-router-imports': modReplaceRouterImports,
};
test('routing steps are run in correct order', async () => {
  const expectedRoutingIds = Object.keys(expectedRoutingSteps);
  const steps = getSteps({version: 14})
    .map(s => s.id)
    .filter(s => {
      return expectedRoutingIds.includes(s);
    });
  expect(steps).toEqual(expectedRoutingIds);
});

test('routing codemods', async () => {
  const contents = fs
    .readFileSync(
      path.join(__dirname, '../__fixtures__/routes-fixture/src/routes.js')
    )
    .toString();
  let transformed = contents;
  Object.values(expectedRoutingSteps).forEach(mod => {
    transformed = babel.transform(transformed, {
      plugins: [mod],
      filename: 'any.js',
      parserOpts,
    }).code;
  });
  transformed = format(transformed);
  expect(transformed).toMatchSnapshot();
});
