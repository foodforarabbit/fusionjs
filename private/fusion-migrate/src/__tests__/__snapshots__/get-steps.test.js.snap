// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`routing codemods 1`] = `
"import {compose} from 'redux';
import {createRPCPromise} from '@uber/web-rpc-redux';
import {connect} from 'react-redux';
import {prepared} from 'fusion-react';
import {withRouter, Redirect, Route, Switch} from 'fusion-plugin-react-router';

const E = args => {
  return createRPCPromise('E', args);
}; // TODO: You may want to move this into the container component file

const _EWithData = compose(
  connect(
    state => {
      // TODO: get things from state you need here
      return {};
    },
    {
      E,
    }
  ),
  prepared(props => {
    // TODO: You probably want to add a check to see if the data exists, or is loading.
    // See https://engdocs.uberinternal.com/web/docs/guides/fetching-data/#use-rpc-method-in-a-component
    return props.E(props.match.params).catch(() => {});
  })
)(E);

const A = args => {
  return createRPCPromise('A', args);
}; // TODO: You may want to move this into the container component file

const _CWithData = compose(
  connect(
    state => {
      // TODO: get things from state you need here
      return {};
    },
    {
      A,
    }
  ),
  prepared(props => {
    // TODO: You probably want to add a check to see if the data exists, or is loading.
    // See https://engdocs.uberinternal.com/web/docs/guides/fetching-data/#use-rpc-method-in-a-component
    return props.A(props.match.params).catch(() => {});
  })
)(C);

const getB = args => {
  return createRPCPromise(rpcIds.getB, args);
};

const getA = args => {
  return createRPCPromise(rpcIds.getA, args);
}; // TODO: You may want to move this into the container component file

const _HomeContainerWithData2 = compose(
  connect(
    state => {
      // TODO: get things from state you need here
      return {};
    },
    {
      getA,
      getB,
    }
  ),
  prepared(props => {
    // TODO: You probably want to add a check to see if the data exists, or is loading.
    // See https://engdocs.uberinternal.com/web/docs/guides/fetching-data/#use-rpc-method-in-a-component
    return Promise.all([
      props[rpcIds.getA](props.match.params),
      props[rpcIds.getB](props.match.params),
    ]).catch(() => {});
  })
)(HomeContainer);

const _routes = (
  <AppContainer
    match={{
      params: {},
    }}
  >
    <Switch>
      <Route
        path=\\"/\\"
        exact
        component={_HomeContainerWithData2}
        trackingId=\\"home\\"
      />
      <Route path=\\"/a\\" component={A} trackingId=\\"a\\" exact />
      <Route
        path=\\"/b\\"
        trackingId=\\"b\\"
        render={_props => {
          return (
            <B {..._props}>
              <Switch>
                <Route
                  path={_props.match.path + '/'}
                  exact
                  render={props => {
                    return <Redirect to=\\"/c\\" />;
                  }}
                />
                <Route
                  path={_props.match.path + '/c'}
                  component={_CWithData}
                  trackingId=\\"c\\"
                  exact
                />
                <Route
                  path={_props.match.path + '/e'}
                  component={_EWithData}
                  trackingId=\\"E\\"
                  exact
                />
                <Route
                  path={_props.match.path + '/d'}
                  exact
                  render={props => {
                    return <Redirect to=\\"/c\\" />;
                  }}
                />
              </Switch>
            </B>
          );
        }}
        exact
      />
    </Switch>
  </AppContainer>
);

export default _routes;
"
`;
